nixos/modules/services/wayland/cage.nix

```diff
@@ -36,8 +36,18 @@ in {
     type = types.path;
     default = "${pkgs.xterm}/bin/xterm";
     defaultText = literalExpression ''"''${pkgs.xterm}/bin/xterm"'';
+    example = literalExpression ''
+      pkgs.writeShellScript "cage-program" '''
+        cd /home/''${user}
```

Comment:
What about using `WorkingDirectory="~"` from systemd instead of changing the working directory manually?  The advantage of my proposal is that we can set `WorkingDirectory="~"` in this module so that users will not encounter this issue any more.

Reply:
I'd also prefer a solution where it'd be solved outright instead of documenting a workaround in an example and hope that users read the docs.

One thing I'm not sure about `WorkingDirectory="~"`, is what happens if no value is passed to `services.cage.user`? Will a home directory be created automatically for the default user (`demo`)?

Reply:
From `man systemd.exec`:

> the specified user and group must have been created statically in the user database no later than the moment the service is started ... If the user does not exist by then program invocation will fail.

Reply:
While `man systemd.exec` does say:

> WorkingDirectory= ... If set to "˜", the home directory of the user specified in User= is used.

But I'm still not sure what would happen in this scenario:

> what happens if no value is passed to `services.cage.user`? Will a home directory be created automatically for the default user ([`demo`](https://github.com/thenbe/nixpkgs/blob/fa67dc68064582b7777f6c60c0acd8b819a067ba/nixos/modules/services/wayland/cage.nix#L12))?

It seems our service definition is based on this upstream [example](https://github.com/cage-kiosk/cage/wiki/Starting-Cage-on-boot-with-systemd), which also does not set working directory. To be honest, I'm hesitant to change the working directory all-together without thoroughly testing it (which I am unable to do at the moment), especially since cage handles permissions/security in a way that I am not familiar with. Perhaps we can document the behavior with this PR, then open a separate PR to address the underlying issue?

Reply:
> One thing I'm not sure about when using WorkingDirectory="~", is what happens if no value is passed to services.cage.user? 

If users do not set `services.cage.user`, then it defaults to `demo`.  Note that the systemd service [runs as that user][1].  So if user `demo` does not exist, the systemd service fails.  In this case, we do not need to consider whether the home directory of user `demo` exists or not.

One thing worth noting is that you do not want `cage` to write things to the [default home directory][2] (`/var/empty`) of system users, which means setting `users.users.<name>.createHome = true` when creating a user for `services.cage.user`.

***

Instead of home directory of user `services.cage.user` which isn't semantically right and may not exist, it is better to
1. set `StateDirectory=cage` to make sure it exist
2. and either `cd` to that directory or set `WorkingDirectory` to that directory.

> Perhaps we can document the behavior with this PR, then open a separate PR to address the underlying issue

Sounds good.


[1]: https://github.com/thenbe/nixpkgs/blob/fa67dc68064582b7777f6c60c0acd8b819a067ba/nixos/modules/services/wayland/cage.nix#L81
[2]: https://github.com/thenbe/nixpkgs/blob/fa67dc68064582b7777f6c60c0acd8b819a067ba/nixos/modules/config/users-groups.nix#L129